File: sdk/go/runner_adapter_test.go
package stigmer

import (
	"context"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// mockRunnerAdapter records all calls for verification.
type mockRunnerAdapter struct {
	sessionsOpened       []string
	sessionsClosed       []string
	executionsCreated    []string
	executionsTerminated []string
}

func (m *mockRunnerAdapter) OnSessionOpened(_ context.Context, sessionID string) error {
	m.sessionsOpened = append(m.sessionsOpened, sessionID)
	return nil
}

func (m *mockRunnerAdapter) OnSessionClosed(_ context.Context, sessionID string) error {
	m.sessionsClosed = append(m.sessionsClosed, sessionID)
	return nil
}

func (m *mockRunnerAdapter) OnWorkflowExecutionCreated(_ context.Context, executionID string) error {
	m.executionsCreated = append(m.executionsCreated, executionID)
	return nil
}

func (m *mockRunnerAdapter) OnWorkflowExecutionTerminated(_ context.Context, executionID string) error {
	m.executionsTerminated = append(m.executionsTerminated, executionID)
	return nil
}

// Compile-time interface satisfaction check.
var _ RunnerAdapter = (*mockRunnerAdapter)(nil)

func TestRunnerAdapter_InterfaceSatisfaction(t *testing.T) {
	// Verify that a struct implementing all methods satisfies the interface.
	var adapter RunnerAdapter = &mockRunnerAdapter{}
	require.NotNil(t, adapter)
}

func TestRunnerAdapter_MockRecordsCalls(t *testing.T) {
	ctx := context.Background()
	adapter := &mockRunnerAdapter{}

	err := adapter.OnSessionOpened(ctx, "ses-1")
	require.NoError(t, err)

	err = adapter.OnSessionOpened(ctx, "ses-2")
	require.NoError(t, err)

	err = adapter.OnSessionClosed(ctx, "ses-1")
	require.NoError(t, err)

	err = adapter.OnWorkflowExecutionCreated(ctx, "wfexec-1")
	require.NoError(t, err)

	err = adapter.OnWorkflowExecutionTerminated(ctx, "wfexec-1")
	require.NoError(t, err)

	assert.Equal(t, []string{"ses-1", "ses-2"}, adapter.sessionsOpened)
	assert.Equal(t, []string{"ses-1"}, adapter.sessionsClosed)
	assert.Equal(t, []string{"wfexec-1"}, adapter.executionsCreated)
	assert.Equal(t, []string{"wfexec-1"}, adapter.executionsTerminated)
}

func TestWithRunnerAdapter_SetsOnClient(t *testing.T) {
	adapter := &mockRunnerAdapter{}

	// Use WithInsecure to avoid needing real credentials.
	client, err := NewClient(
		WithBaseURL("localhost:7234"),
		WithInsecure(),
		WithRunnerAdapter(adapter),
	)
	require.NoError(t, err)
	defer client.Close()

	assert.Equal(t, adapter, client.RunnerAdapter)
}

func TestWithRunnerAdapter_NilIsNoOp(t *testing.T) {
	client, err := NewClient(
		WithBaseURL("localhost:7234"),
		WithInsecure(),
	)
	require.NoError(t, err)
	defer client.Close()

	assert.Nil(t, client.RunnerAdapter)
}

File: sdk/java/src/test/java/ai/stigmer/sdk/RunnerAdapterTest.java
package ai.stigmer.sdk;

import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

class RunnerAdapterTest {

    /** Mock adapter that records all lifecycle calls. */
    static class MockRunnerAdapter implements RunnerAdapter {
        final List<String> sessionsOpened = new ArrayList<>();
        final List<String> sessionsClosed = new ArrayList<>();
        final List<String> executionsCreated = new ArrayList<>();
        final List<String> executionsTerminated = new ArrayList<>();

        @Override
        public void onSessionOpened(String sessionId) {
            sessionsOpened.add(sessionId);
        }

        @Override
        public void onSessionClosed(String sessionId) {
            sessionsClosed.add(sessionId);
        }

        @Override
        public void onWorkflowExecutionCreated(String executionId) {
            executionsCreated.add(executionId);
        }

        @Override
        public void onWorkflowExecutionTerminated(String executionId) {
            executionsTerminated.add(executionId);
        }
    }

    @Test
    void mockAdapter_recordsCalls() throws Exception {
        MockRunnerAdapter adapter = new MockRunnerAdapter();

        adapter.onSessionOpened("ses-1");
        adapter.onSessionOpened("ses-2");
        adapter.onSessionClosed("ses-1");
        adapter.onWorkflowExecutionCreated("wfexec-1");
        adapter.onWorkflowExecutionTerminated("wfexec-1");

        assertEquals(List.of("ses-1", "ses-2"), adapter.sessionsOpened);
        assertEquals(List.of("ses-1"), adapter.sessionsClosed);
        assertEquals(List.of("wfexec-1"), adapter.executionsCreated);
        assertEquals(List.of("wfexec-1"), adapter.executionsTerminated);
    }

    @Test
    void builder_withRunnerAdapter_setsAdapter() {
        MockRunnerAdapter adapter = new MockRunnerAdapter();

        try (StigmerClient client = StigmerClient.builder("sk_test_key")
                .runnerAdapter(adapter)
                .build()) {
            assertSame(adapter, client.runnerAdapter());
        }
    }

    @Test
    void builder_withoutRunnerAdapter_returnsNull() {
        try (StigmerClient client = StigmerClient.builder("sk_test_key").build()) {
            assertNull(client.runnerAdapter());
        }
    }

    @Test
    void interface_isImplementable() {
        RunnerAdapter adapter = new RunnerAdapter() {
            @Override
            public void onSessionOpened(String sessionId) {}
            @Override
            public void onSessionClosed(String sessionId) {}
            @Override
            public void onWorkflowExecutionCreated(String executionId) {}
            @Override
            public void onWorkflowExecutionTerminated(String executionId) {}
        };

        assertDoesNotThrow(() -> {
            adapter.onSessionOpened("test");
            adapter.onSessionClosed("test");
            adapter.onWorkflowExecutionCreated("test");
            adapter.onWorkflowExecutionTerminated("test");
        });
    }
}
